Enhance: allow profiles to be reordered when multi-playing

SlySven opened the pull request:
This implements a requested feature to allow the tabs on the tab bar to be reordered and the profiles displayed underneath when multi-playing to follow suite. This not being done previously was why I had to lock the tab bar because at that time the tabs could move but the profiles did not!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight

~~Tabs and their associated profiles' Main Consoles can now be rearranged when multi-playing.~~
Profiles can now be rearranged when multi-playing.

SlySven committed:
Enhance: allow profiles to be reordered when multi-playing

This implements a requested feature to allow the tabs on the tab bar to
reordered and the profiles displayed underneath when multi-playing will
follow suite. This not being done previously was why I had to lock the
tab bar because at that time the tabs could move but the profiles did not!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Cleanup: remove unused function that is impacted by this PR

I *knew* there was a reason why I had removed it - but neglected to include
that in the original commit.

Also update some of my Copyright lines.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> Tabs and their associated profiles' Main Consoles can now be rearranged when multi-playing.

Profiles can now be rearranged when multi-playing.

SlySven committed:
Revise: remove static const TMainConsole::scmProperty_HostName

Given that we only use *this* instance of the `const char "HostNme"` in
THREE places in THREE separate files it was felt not to be worthwhile
abstracting it to an explicit `const` value and just to use the raw literal
where needed. As it happens the same text is used elsewhere for the same
purpose (identifying the related profile in a Qt rather than a Mudlet class
instance) there might be a more optimum way to do this in a common
universally used (maybe even pre-compiled) header file.

However that can be left for a separate PR if it is thought worthwhile...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: please clang-tidy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/mudlet.cpp:
```diff
@@ -4331,3 +4323,35 @@ void mudlet::setupTrayIcon()
+    // gone through them all it will mean that they are in the same order as the
+    // tabs:
+    for (int index = 0; index < itemsCount; ++index) {
+        auto wantedTabName = tabNamesInOrder.at(index);
```
Done.

vadi2 commented on src/TMainConsole.cpp:
```diff
@@ -74,6 +77,13 @@ TMainConsole::TMainConsole(Host* pH, QWidget* parent)
     // Mudlet version:
     setProfileSpellDictionary();
+
+    // Ensure the QWidget has the profile name embedded into it - since
+    // TMainConsole is derived from TConsole which is derived from QWidget which
+    // is derived from QObject then adding something to the
+    // (QList<QByteArray>) QObject::dynamicPropertyNames should be visible from
+    // the QWidget:
```
```suggestion
    // Ensure the QWidget has the profile name embedded into it
```

:innocent:

SlySven replied:
The use of `(bool) QObject::setProperty(const char *name, const QVariant &value)` (and it's companion getter) is something that I have sneaked into the Mudlet code over the last couple of years but I was not sure if anyone else *gets it* - so I was trying to explain it!

vadi2 commented on src/TMainConsole.h:
```diff
@@ -103,6 +103,8 @@ class TMainConsole : public TConsole
     bool loadMap(const QString&);
     bool importMap(const QString&, QString* errMsg = Q_NULLPTR);
 
+    static const char* scmProperty_HostName;
```
Suggestion - instead of having these properties - why don't we just implement a `getName` or something similar method? This will be more intuitive to use and will show up in autocompletion as well.

SlySven replied:
Because you would still need to use them to embed the profile name into the base (`QObject`) class **unless** the thing returned from `mpHBoxLayout_profileContainer->itemAt(profileIndex)->widget()`, which seems to be a `QWidget` and NOT the `TConsole` or `TMainConsole` derived class can be cast to the derived class with `dynamic_cast<TMainConsole*>()` as https://stackoverflow.com/q/18873871/4805858 suggests - but I'm not sure about the requirement for there to be at least one `virtual function`.

I do not need a `TMainConsole*` here - I just need to know which `QLayoutItem*` points to the top level *container* for which profile - so embedding that information in the base `QObject` that each item is derived from works fine for this.

vadi2 replied:
Yes it does, but how can get get rid of unintuitive things like `scmProperty_HostName`? Do we really need it? I don't think so because it's set in 1 place and read in 1 other.

SlySven replied:
Well either we declare it in an obvious place (at the top of the compilation unit) - or we have a raw literal string `"HostName"` (though not one that is converted to a `QString`!) used in two places where they must be the same.

As it happens I have already used `scmProperty_` as a prefix on some constants elsewhere (see `TTimer::scmProperty_HostName` and `TTimer::scmProperty_TTimerId`) for the same sort of thing {adding meta-data to a `QObject` derived class} and the meta-properties that are involved are accessed in the `mudlet` and `TimerUnit` classes as well as the `TTimer` one. To be fair it would be nicer not to have to tie the definitions to a particular class - in this case the `TMainConsole` one but it does, I suppose say where to go to look for the raw string being used - even though the whole idea of using a ***constant*** instead of the same literal in different places, **even** though it is only **two** for this one, is to abstract that away. Of course `TTimer::scmProperty_HostName` and `TMainConsole::scmProperty_HostName` also have the same **value** but they are used in different places to store the same piece of data in different `QObject`-derived class instances: `TTimer` and `TMainConsole` ones respectively.

tl;dr; Are you insisting that I replace the uses of `TMainConsole::scmProperty_HostName` with `"HostName"`?

vadi2 replied:
Yeah. Abstractions, just like any tool, are best used sparingly and in this case it's adding more trouble than help. Something like `TMainConsole` vs `TConsole` - good idea, makes sense - but this is too little and the price is too much.

SlySven replied:
Done in https://github.com/Mudlet/Mudlet/pull/4976/commits/6405a56e8c119b57366cb14d3aadced82a87646f.

vadi2 reviewed (commented):
This is wonderful magic and it works amazing.

SlySven committed:
Update src/TMainConsole.cpp

Considerably shorten an attempt to explain what is going on.

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

Kebap commented:
This seems to change menu and icon bar fonts from release version:
![image](https://user-images.githubusercontent.com/117238/111542683-55c61b00-8772-11eb-891d-f88aca2040f4.png)

Kebap reviewed (approved):
Easy to use!

SlySven commented:
TBH That looks like the release version is using `"Bitstream Vera Sans Mono"` as the `(QFont) mainFont` in the `mudlet` constructor - as opposed to `"Bitstream Vera Sans"` in the test build for this PR. Personally I think the latter actually looks better in the GUI (*IMHO*) and I have made a tweak in that area in some local branches - but I can't see that I have pushed out any PR with that change in...

As far as I can tell the font for the main toolbar and main menu is derived from the default application font in the absence of a font being set for the `mudlet` class when they are created - but neither a `mudlet` class one or the application one has been set at that point so will be whatever Qt configures it to be...

~~Do you have a preference yourself, @Kebap ?~~ *Edit: I do not think this is something that this PR needs to address.*

vadi2 reviewed (approved):
Thanks a lot! This is an awesome improvement.

SlySven commented:
> This seems to change menu and icon bar fonts from release version:
> ![image](https://user-images.githubusercontent.com/117238/111542683-55c61b00-8772-11eb-891d-f88aca2040f4.png)

Actually after a decent thunk I've realised that for the main application window (and it's GUI) we should not really be specifying a font to use - instead we should be making use of whatever `QFont()`, ***without*** a font/filename, returns. This will be the default font set by the Desktop Environment (if Qt can work it out) or a *reasonable* (for some select value of reasonable!) default if the Qt libraries can't. We can probably adjust the font *size* but other details should be left as they are so that Mudlet adopts the same appearance (as far as possible) as other Qt applications in the same Desktop Environment.

Of course this will NOT extend to things like *AppImage* based builds because they use self contained Qt Libraries which may (probably won't) interact with the DE/OS to the same degree as system Qt libraries; indeed I wonder whether that explains the differences you are seeing in those two builds...
